### PR TITLE
[Traders] Fix legacy migration

### PR DESCRIPTION
**Pull request by Herklos**
requires #25

### CODE REVIEW

**Comment by Herklos** [comment]
Too big, I'll split into multiple commits.

**Comment by Herklos** [comment]
Splitted into 3 commits :+1:

File: octobot_trading/data/order.py [hunk 1]
```diff
@@ -254,38 +264,67 @@ def is_self_managed(self):
...
         if self.side is None or self.order_type is None:
-            self.__update_type_from_raw(raw_order)
-            if self.taker_or_maker is None:
-                self.__update_taker_maker_from_raw()
+            try:
+                self.__update_type_from_raw(raw_order)
+                if self.taker_or_maker is None:
+                    self.__update_taker_maker_from_raw()
+            except KeyError:
+                pass
```
Comment (GuillaumeDSM) [inline octobot_trading/data/order.py]: Is it really good to silently pass on an order parsing error ?

File: octobot_trading/data/order.py [hunk 2]
```diff
@@ -254,38 +264,67 @@ def is_self_managed(self):
...
+            "filled_price": get_value_or_default(raw_order, ExchangeConstantsOrderColumns.PRICE.value, None),
+            "total_cost": get_value_or_default(raw_order, ExchangeConstantsOrderColumns.COST.value, None),
+            "fee": get_value_or_default(raw_order, ExchangeConstantsOrderColumns.FEE.value, None),
+            "timestamp": get_value_or_default(raw_order, ExchangeConstantsOrderColumns.TIMESTAMP.value, None)
         })
 
-    def __update_type_from_raw(self, raw_order):
-        self.side: TradeOrderSide = TradeOrderSide(raw_order[ExchangeConstantsOrderColumns.SIDE.value])
+    @staticmethod
+    def parse_order_type(raw_order):
```
Comment (GuillaumeDSM) [inline octobot_trading/data/order.py]: I guess it's a good time to integrate new order types from https://github.com/Drakkar-Software/OctoBot/commit/3160e197bd841a49e2e9b54a43bf4fa13f03bb02

File: octobot_trading/data/order.py [hunk 3]
```diff
@@ -254,38 +264,67 @@ def is_self_managed(self):
...
-                self.order_type = TraderOrderType.SELL_MARKET
+                order_type = TraderOrderType.SELL_MARKET
+        return side, order_type
+
+    @staticmethod
+    def parse_order_status(raw_order):
+        try:
+            return OrderStatus(raw_order[ExchangeConstantsOrderColumns.STATUS.value])
+        except KeyError:
+            return None
```
Comment (GuillaumeDSM) [inline octobot_trading/data/order.py]: Same as above, it a status is not recognized, I think it's better not to silently pass

File: tests/orders/test_order_factory.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+from octobot_trading.data.order import Order
+from octobot_trading.enums import TradeOrderSide, TradeOrderType, TraderOrderType
+from octobot_trading.exchanges.exchange_manager import ExchangeManager
+from octobot_trading.traders.trader_simulator import TraderSimulator
+
+pytestmark = pytest.mark.asyncio
+
+
+class TestOrderFactory:
```
Comment (GuillaumeDSM) [inline tests/orders/test_order_factory.py]: :+1:

File: octobot_trading/orders/order_factory.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+#  Drakkar-Software OctoBot-Trading
```
Comment (GuillaumeDSM) [inline octobot_trading/orders/order_factory.py]: It looks like methods from this file are not used yet, I guess it's for futur dependencies ?

File: octobot_trading/orders/order_util.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    current_symbol_holding = exchange_manager.exchange_personal_data.portfolio_manager.portfolio\
+        .get_currency_portfolio(currency)
+    current_market_quantity = exchange_manager.exchange_personal_data.portfolio_manager.portfolio\
+        .get_currency_portfolio(market)
+
+    market_quantity = current_market_quantity / mark_price
+
+    symbol_market = exchange_manager.exchange.get_market_status(symbol, with_fixer=False)
+
+    return current_symbol_holding, current_market_quantity, market_quantity, mark_price, symbol_market
```
Comment (GuillaumeDSM) [inline octobot_trading/orders/order_util.py]: :+1:  for splitting methods sementically

File: octobot_trading/consumers/abstract_mode_consumer.py [hunk 7]
```diff
@@ -25,30 +23,18 @@
...
-        self.logger = get_logger(self.__class__.__name__)
+        self._logger = get_logger(self.__class__.__name__)
 
         self.trading_mode = trading_mode
         self.exchange_manager = trading_mode.exchange_manager
-        self.trader = self.exchange_manager.trader
-
-        # shortcuts
-        self.exchange_personal_data = self.exchange_manager.exchange_personal_data
-        self.exchange_portfolio_manager = self.exchange_personal_data.portfolio_manager
```
Comment (GuillaumeDSM) [inline octobot_trading/consumers/abstract_mode_consumer.py]: :tornado: :+1:

File: tests/exchanges/test_exchanges.py [hunk 8]
```diff
@@ -74,3 +72,36 @@ class TestExchanges:
 
         with pytest.raises(KeyError):
             assert Exchanges.instance().get_exchange("test")
+
+        await exchange_manager_binance.stop()
+        await exchange_manager_bitmex.stop()
+        await exchange_manager_poloniex.stop()
+
+    async def test_del_exchange(self):
```
Comment (GuillaumeDSM) [inline tests/exchanges/test_exchanges.py]: :+1:

Comment (GuillaumeDSM) [inline tests/exchanges/test_exchanges.py]: :+1:

File: octobot_trading/consumers/abstract_mode_consumer.py (hunk 7 above)
Comment (GuillaumeDSM) [inline octobot_trading/consumers/abstract_mode_consumer.py]: :tornado: :+1:

File: octobot_trading/orders/order_util.py (hunk 6 above)
Comment (GuillaumeDSM) [inline octobot_trading/orders/order_util.py]: great to split this utility methods into their file

File: octobot_trading/exchanges/exchange_manager.py [hunk 9]
```diff
@@ -109,42 +115,51 @@ def reset_exchange_personal_data(self):
...
             await self._create_exchange_producers()
 
         if self.is_backtesting:
-            try:
-                await self.exchange.modify_channels()
-                await self.exchange.create_backtesting_exchange_producers()
-            except ValueError:
-                self.logger.error("Not enough exchange data to calculate backtesting duration")
-                await self.stop()
+            await self._init_simulated_exchange()
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/exchange_manager.py]: much clearer :100:

File: tests/exchanges/data/test_exchange_config_data.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+        assert "ICX/BTC" in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        assert "NEO/BTC" in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        assert "VEN/BTC" in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        assert "XLM/BTC" in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        assert "ONT/BTC" in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        assert "BTC/USDT" in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        assert "ETH/USDT" not in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        assert "NEO/BNB" not in exchange_manager.exchange_config.traded_crypto_currencies_pairs["Bitcoin"]
+        await exchange_manager.stop()
```
Comment (GuillaumeDSM) [inline tests/exchanges/data/test_exchange_config_data.py]: :+1:

**Comment by GuillaumeDSM** [comment]
A bit too big on an analysis from me, I didn't realise other PRs share some commits, attention requiring comments should be about this PR though

**Comment by Herklos** [comment]
> A bit too big on an analysis from me, I didn't realise other PRs share some commits, attention requiring comments should be about this PR though

Yeah, I split into multiple commit to simplify the analysis and add the requirement with the #25, so you should have review this PR and its requirements first.

File: octobot_trading/data/order.py (hunk 2 above)
Comment (Herklos) [reply octobot_trading/data/order.py]: The new order type have already been integrated in this folder : https://github.com/Drakkar-Software/OctoBot-Trading/tree/master/octobot_trading/orders (trailing stop etc...). Are you talking about that ?

File: octobot_trading/orders/order_factory.py (hunk 5 above)
Comment (Herklos) [reply octobot_trading/orders/order_factory.py]: For example create_order_instance is used in TradingMode before calling trader.create_order.

File: octobot_trading/data/order.py (hunk 1 above)
Comment (Herklos) [reply octobot_trading/data/order.py]: I agree :+1:

File: octobot_trading/data/order.py (hunk 3 above)
Comment (Herklos) [reply octobot_trading/data/order.py]: I agree (should raise something) :+1:

**Comment by GuillaumeDSM** [comment]
> > A bit too big on an analysis from me, I didn't realise other PRs share some commits, attention requiring comments should be about this PR though
> 
> Yeah, I split into multiple commit to simplify the analysis and add the requirement with the #25, so you should have review this PR and its requirements first.

Right

File: octobot_trading/data/order.py (hunk 2 above)
Comment (GuillaumeDSM) [reply octobot_trading/data/order.py]: Right, but to my understanding, if we have a stop profit type order from exchange for example , it won't be recognized in this method (ex if order_type == TradeOrderType.TAKE_PROFIT). Or maybe you mean it's not a problem ?

Comment (Herklos) [reply octobot_trading/data/order.py]: Oh yeah you are right :+1:

Comment (Herklos) [reply octobot_trading/data/order.py]: Actually the problem come from CCXT. CCXT is giving us the side of the order (BUY or SELL) and a MARKET or a LIMIT. There is currently no possibility to have something else. Except for private methods implemented in specific exchanges that shouldn't be supported here I think.

Comment (GuillaumeDSM) [reply octobot_trading/data/order.py]: The code in https://github.com/Drakkar-Software/OctoBot/commit/3160e197bd841a49e2e9b54a43bf4fa13f03bb02 ex: ```order_type == TradeOrderType.TAKE_PROFIT_LIMIT``` was working for me with binance new order types with real trader, without any specific code, maybe it's not broken ?

File: octobot_trading/data/order.pxd [hunk 11]
```diff
@@ -13,7 +13,7 @@
 #  Lesser General License for more details.
 #
 #  You should have received a copy of the GNU Lesser General Public
-#  License along with this library.
+#  License adouble with this library.
```
Comment (GuillaumeDSM) [inline octobot_trading/data/order.pxd]: 😂

File: octobot_trading/data/order.py [hunk 12]
```diff
@@ -271,60 +271,36 @@ def update_from_raw(self, raw_order):
...
-        self.status = Order.parse_order_status(raw_order)
+        self.status = parse_order_status(raw_order)
         self.total_cost = raw_order[ExchangeConstantsOrderColumns.COST.value]
         self.filled_quantity = raw_order[ExchangeConstantsOrderColumns.FILLED.value]
         self.filled_price = raw_order[ExchangeConstantsOrderColumns.PRICE.value]
         if not self.filled_price and self.filled_quantity:
             self.filled_price = self.total_cost / self.filled_quantity
-        self.taker_or_maker = Order.parse_order_type(raw_order)
+
+        self.taker_or_maker = parse_order_type(raw_order)
```
Comment (GuillaumeDSM) [inline octobot_trading/data/order.py]: Is parse_order_type(raw_order) telling if the order is a taker or a maker ?

File: octobot_trading/exchanges/rest_exchange.py [hunk 13]
```diff
@@ -359,11 +358,19 @@ def get_uniform_timestamp(self, timestamp):
         self.logger.info(f"Connection closed.")
 
     def get_pair_from_exchange(self, pair) -> str:
-        return self.client.market(pair)["symbol"]
+        try:
+            return self.client.market(pair)["symbol"]
+        except ccxt.base.errors.BadSymbol:
+            self.logger.error(f"Failed to get market of {pair}")
+            return None
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/rest_exchange.py]: 👍

File: octobot_trading/exchanges/rest_exchange.py [hunk 14]
```diff
@@ -359,11 +358,19 @@ def get_uniform_timestamp(self, timestamp):
...
 
     def get_split_pair_from_exchange(self, pair) -> (str, str):
-        market_data: dict = self.client.market(pair)
-        return market_data["base"], market_data["quote"]
+        try:
+            market_data: dict = self.client.market(pair)
+            return market_data["base"], market_data["quote"]
+        except BadSymbol:
+            self.logger.error(f"Failed to get market of {pair}")
+            return None, None
```
Comment (GuillaumeDSM) [inline octobot_trading/exchanges/rest_exchange.py]: 👍

File: tests/traders/test_trader.py [hunk 15]
```diff
@@ -483,21 +484,23 @@ class TestTrader:
             PORTFOLIO_TOTAL: 1000
         }
 
-        with patch.object(PricesManager, 'get_mark_price', new=AsyncMock(return_value=1)):
-            orders = await trader_inst.sell_all()
+        if not os.getenv('CYTHON_TEST_IGNORE'):
```
Comment (GuillaumeDSM) [inline tests/traders/test_trader.py]: sad to have to do this but if no choice then I guess it's best this way

File: octobot_trading/data/order.pxd (hunk 11 above)
Comment (Herklos) [reply octobot_trading/data/order.pxd]: Hum :laughing:

File: tests/traders/test_trader.py (hunk 15 above)
Comment (Herklos) [reply tests/traders/test_trader.py]: Yes I tested may ways to avoid doing this...

File: octobot_trading/data/order.py (hunk 12 above)
Comment (Herklos) [reply octobot_trading/data/order.py]: Oh no you're right, I don't know where this line come from.

File: octobot_trading/data/order.py (hunk 1 above)
Comment (GuillaumeDSM) [reply octobot_trading/data/order.py]: Can you add an error log or a raise ? 😄

Comment (Herklos) [reply octobot_trading/data/order.py]: Yes sorry ^^

**Pull request merged**